File: src/main/java/jp/kentan/minecraft/neko_core/component/zone/Area.java
package jp.kentan.minecraft.neko_core.component.zone;

import jp.kentan.minecraft.neko_core.util.Log;
import jp.kentan.minecraft.neko_core.util.Util;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.OfflinePlayer;
import org.bukkit.block.BlockState;
import org.bukkit.block.Sign;

import java.time.ZonedDateTime;
import java.util.UUID;

public class Area {

    private final static String SIGN_INDEX_TEXT = ChatColor.DARK_GRAY + "" + ChatColor.BOLD + "[" + ChatColor.BLUE + ChatColor.BOLD + "区画" + ChatColor.DARK_GRAY + ChatColor.BOLD + "]";

    public final String NAME, WORLD, ZONE_ID, REGION_ID;
    public final int REGION_SIZE;
    public final AreaState STATE;
    public final Location SIGN_LOCATION;
    public final double BUY_RENTAL_PRICE;
    public final ZonedDateTime EXPIRE_DATE;

    public final OfflinePlayer OWNER;

    public Area(
            String name,
            String world,
            String zoneId,
            String regionId,
            int regionSize,
            AreaState state,
            Location signLocation,
            UUID ownerUuid,
            double buyRentalPrice,
            ZonedDateTime expireDate) {
        NAME = name;
        WORLD = world;
        ZONE_ID = zoneId;
        REGION_ID = regionId;
        REGION_SIZE = regionSize;
        STATE = state;
        SIGN_LOCATION = signLocation;
        BUY_RENTAL_PRICE = buyRentalPrice;
        EXPIRE_DATE = expireDate;

        OWNER = (ownerUuid != null) ? Bukkit.getOfflinePlayer(ownerUuid) : null;
    }

    public boolean isOwner(UUID uuid) {
        return (OWNER != null) && OWNER.getUniqueId().equals(uuid);
    }

    public final boolean isLock() {
        return STATE == AreaState.LOCK;
    }

    public final boolean onSale() {
        return STATE == AreaState.ON_SALE;
    }

    public boolean updateSign(){
        if(SIGN_LOCATION == null){
            return true;
        }

        BlockState blockState = SIGN_LOCATION.getWorld().getBlockAt(SIGN_LOCATION).getState();

        if(blockState instanceof Sign){
            Sign sign = (Sign)blockState;

            sign.setLine(0, SIGN_INDEX_TEXT);

            if(OWNER != null){
                sign.setLine(2, ChatColor.DARK_GRAY + OWNER.getName());
                sign.setLine(3, (EXPIRE_DATE == null) ? STATE.getName() : (ChatColor.DARK_PURPLE + Util.formatDate(EXPIRE_DATE)) + ChatColor.RESET + " まで");
            }else {
                sign.setLine(2, "");
                sign.setLine(3, STATE.getName());
            }

            if(!sign.update()){
                Log.warn("failed to update Area sing at " +
                        SIGN_LOCATION.getX() + ", " +
                        SIGN_LOCATION.getY() + ", " +
                        SIGN_LOCATION.getZ() + " in " + SIGN_LOCATION.getWorld().getName());
            }
        }else{ //看板が見つからなかった場合
            Log.warn("Sign was not found at " +
                    SIGN_LOCATION.getX() + ", " +
                    SIGN_LOCATION.getY() + ", " +
                    SIGN_LOCATION.getZ() + " in " + SIGN_LOCATION.getWorld().getName());

            return false;
        }

        return true;
    }
}

File: src/main/kotlin/jp/kentan/minecraft/nekocore/data/model/Area.kt
package jp.kentan.minecraft.nekocore.data.model

import jp.kentan.minecraft.nekocore.util.formatYearMonthDayHm
import org.bukkit.Location
import org.bukkit.OfflinePlayer
import org.bukkit.World
import org.bukkit.block.Sign
import org.bukkit.entity.Player
import java.util.*

data class Area(
    val id: Int,
    val name: String,
    val world: World,
    val zoneId: String,
    val regionId: String,
    val regionSize: Int,
    val state: State,
    val signLocation: Location?,
    val owner: OfflinePlayer?,
    val purchasedPrice: Double, // ownerが購入・レンタルに支払った金額
    val expiredDate: Date?
) {

    companion object {
        const val SIGN_INDEX_TEXT = "§8§l[§9§l区画§8§l]"
    }

    enum class State(
        val displayName: String
    ) {
        ON_SALE("§c販売中"),
        SOLD("§6契約済み"),
        LOCK("§8ロック中");
    }

    fun isOwner(player: Player) = owner != null && owner.uniqueId == player.uniqueId

    fun updateSign(): Boolean {
        if (signLocation == null) {
            return true
        }

        val sign = signLocation.block.state as? Sign ?: return false

        sign.setLine(0, SIGN_INDEX_TEXT)
        sign.setLine(1, name)
        if (owner != null) {
            sign.setLine(2, "§8${owner.name}")
            sign.setLine(
                3,
                if (expiredDate == null) state.displayName else "§5${expiredDate.formatYearMonthDayHm()}§r まで"
            )
        } else {
            sign.setLine(2, "")
            sign.setLine(3, state.displayName)
        }

        return sign.update()
    }
}